File: web/src/lib/scheme.ts
// Shapes, totals, matching and validation for an assignment's question list and its mark scheme / rubric.
// Mirrors src/sms/schemas/scheme.py (q_label, scheme_total) so the editor shows what the server will store.
import type { Band, Criterion, MarkPoint, MarkSchemeEntry, Question, Rubric, RubricBands, SchemeKind } from "../api/types";
import { rowsToRubricJson, validateRows, type Row } from "./rubric";

export type { Band, MarkPoint, MarkSchemeEntry, Question, RubricBands, SchemeKind };
export type Scheme = MarkSchemeEntry[] | RubricBands[];

export const kindLabel: Record<SchemeKind, string> = { mark_scheme: "Maths / Science — mark scheme", rubric: "Essay — rubric", criteria: "Quick mark" };

export const emptyQuestion = (): Question => ({ q_id: "", text: "", max_marks: 1 });
export const emptySchemeRow = (q_id = ""): MarkSchemeEntry => ({ q_id, answer: "", marks: [], notes: "" });
export const emptyMarkPoint = (): MarkPoint => ({ label: "", marks: 1 });
export const emptyCriterion = (): RubricBands => ({ criterion: "", bands: [emptyBand()] });
export const emptyBand = (): Band => ({ band: "", marks: 1, descriptor: "" });

const Q_ID = /^(q)?(\d+)([a-z])?([ivx]+)?$/i;
const ROMAN = /^[ivx]+$/i;

/** Teacher-facing label for a question id: "1a" → "1(a)", "2bii" → "2(b)(ii)", "4iii" → "4(iii)", "q1" → "Q1".
 *  Anything else (already formatted, free text) is returned unchanged. */
export function qLabel(qId: string): string {
  const raw = (qId ?? "").trim();
  const m = Q_ID.exec(raw);
  if (!m) return raw;
  const [, prefix, number, letter, roman] = m;
  const suffix = (letter ?? "") + (roman ?? "");
  const parts = suffix.length > 1 && ROMAN.test(suffix)
    ? [suffix.toLowerCase()] // "4iii": one roman part, not the letter "i" plus "ii"
    : [letter, roman].filter((p): p is string => !!p).map((p) => p.toLowerCase());
  return (prefix ? "Q" : "") + number + parts.map((p) => `(${p})`).join("");
}

const int = (v: unknown) => (Number.isInteger(Number(v)) ? Number(v) : 0);
const key = (id: string) => (id ?? "").trim();

export const rowTotal = (row: MarkSchemeEntry): number => (row.marks ?? []).reduce((s, m) => s + int(m.marks), 0);
export const bestBand = (c: RubricBands): number => (c.bands ?? []).reduce((s, b) => Math.max(s, int(b.marks)), 0);

/** Total marks the assignment is out of — same rule as the server's scheme_total. */
export function schemeTotal(kind: SchemeKind, questions: Question[], scheme: Scheme): number {
  if (kind === "rubric") return (scheme as RubricBands[]).reduce((s, c) => s + bestBand(c), 0);
  if (kind === "mark_scheme") {
    const rows = new Map<string, number>();
    for (const r of scheme as MarkSchemeEntry[]) if (!rows.has(key(r.q_id))) rows.set(key(r.q_id), rowTotal(r));
    if (questions.length === 0) return Array.from(rows.values()).reduce((s, v) => s + v, 0);
    return questions.reduce((s, q) => s + (rows.get(key(q.q_id)) ?? int(q.max_marks)), 0);
  }
  return questions.reduce((s, q) => s + int(q.max_marks), 0);
}

/** q_ids of questions that have no mark-scheme row, in paper order. */
export function unmatchedQuestions(questions: Question[], scheme: MarkSchemeEntry[]): string[] {
  const rows = new Set(scheme.map((r) => key(r.q_id)));
  return questions.map((q) => key(q.q_id)).filter((id) => !rows.has(id));
}

/** Indexes of mark-scheme rows whose q_id is not on the paper (blank ids included). */
export function orphanRows(questions: Question[], scheme: MarkSchemeEntry[]): number[] {
  const ids = new Set(questions.map((q) => key(q.q_id)));
  return scheme.map((r, i) => (ids.has(key(r.q_id)) ? -1 : i)).filter((i) => i >= 0);
}

function listLabels(ids: string[]): string {
  const labels = ids.map(qLabel);
  if (labels.length === 1) return labels[0];
  if (labels.length <= 3) return `${labels.slice(0, -1).join(", ")} and ${labels[labels.length - 1]}`;
  return `${labels.slice(0, 3).join(", ")} and ${labels.length - 3} more`;
}

function validateQuestions(questions: Question[]): string | null {
  const seen = new Set<string>();
  for (const q of questions) {
    const id = key(q.q_id);
    if (!id) return "Every question needs an id, like 1a.";
    if (seen.has(id)) return `Question ids must be unique — ${qLabel(id)} appears twice.`;
    seen.add(id);
    if (!Number.isInteger(Number(q.max_marks)) || Number(q.max_marks) < 0) return `Marks for ${qLabel(id)} must be a whole number of 0 or more.`;
  }
  return null;
}

/** Why the assignment cannot be saved yet, in one plain sentence — or null when it can. */
export function validateTemplate(kind: SchemeKind, questions: Question[], scheme: Scheme, criteria: Row[]): string | null {
  if (kind === "criteria") return validateRows(criteria);
  if (questions.length === 0) return "Add at least one question.";
  const qProblem = validateQuestions(questions);
  if (qProblem) return qProblem;
  if (kind === "mark_scheme") {
    for (const r of scheme as MarkSchemeEntry[]) {
      const id = key(r.q_id);
      if (!id) return "Every scheme row needs a question id.";
      for (const m of r.marks ?? []) {
        if (!key(m.label)) return `Every allocation for ${qLabel(id)} needs a label, like M1.`;
        if (!Number.isInteger(Number(m.marks)) || Number(m.marks) < 0) return `Allocation marks for ${qLabel(id)} must be a whole number of 0 or more.`;
      }
    }
    const missing = unmatchedQuestions(questions, scheme as MarkSchemeEntry[]);
    if (missing.length === 1) return `Add a scheme row for ${qLabel(missing[0])}.`;
    if (missing.length > 1) return `Add scheme rows for ${listLabels(missing)}.`;
    return null;
  }
  const rows = scheme as RubricBands[];
  if (rows.length === 0) return "Add at least one criterion.";
  for (const c of rows) {
    const name = key(c.criterion);
    if (!name) return "Every criterion needs a name.";
    if (!c.bands || c.bands.length === 0) return `Add at least one band to ${name}.`;
    for (const b of c.bands) {
      if (!key(b.band)) return `Every band for ${name} needs a name.`;
      if (!Number.isInteger(Number(b.marks)) || Number(b.marks) < 0) return `Band marks for ${name} must be a whole number of 0 or more.`;
    }
  }
  return null;
}

const slug = (s: string) => s.trim().toLowerCase().replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "");

function uniqueIds(defs: Criterion[]): Criterion[] {
  const seen = new Map<string, number>();
  return defs.map((d) => {
    const base = d.id || "c";
    const n = (seen.get(base) ?? 0) + 1;
    seen.set(base, n);
    return { ...d, id: n === 1 ? base : `${base}-${n}` };
  });
}

/** The batch-2 `rubric` the API still requires for every kind: for a mark scheme one criterion per question
 *  (worth its scheme row, or the paper's marks), for a rubric one per criterion (worth its best band). An empty
 *  draft gets a single zero-mark placeholder so it can be stored before the paper is read. */
export function placeholderRubric(kind: SchemeKind, questions: Question[], scheme: Scheme, criteria: Row[]): Rubric {
  let defs: Criterion[];
  if (kind === "criteria") {
    defs = criteria.length ? (JSON.parse(rowsToRubricJson(criteria)) as Rubric).criterion_defs : [];
  } else if (kind === "mark_scheme") {
    const rows = new Map<string, number>();
    for (const r of scheme as MarkSchemeEntry[]) if (!rows.has(key(r.q_id))) rows.set(key(r.q_id), rowTotal(r));
    defs = questions.filter((q) => key(q.q_id)).map((q) => ({
      id: key(q.q_id), description: (q.text || "").trim().slice(0, 200) || qLabel(q.q_id),
      max_score: rows.get(key(q.q_id)) ?? int(q.max_marks),
    }));
  } else {
    defs = (scheme as RubricBands[]).filter((c) => key(c.criterion)).map((c) => ({
      id: slug(c.criterion) || "c", description: c.criterion.trim(), max_score: bestBand(c),
    }));
  }
  if (defs.length === 0) defs = [{ id: "draft", description: "Draft", max_score: 0 }];
  return { criterion_defs: uniqueIds(defs) };
}

File: src/sms/schemas/scheme.py
"""Canonical shapes of an assignment's question list and its mark scheme / rubric — the JSON stored
in assignment_templates.questions_json and scheme_json and the models the extraction agents fill in."""
import re
from typing import Any, Iterable, List, Union

from pydantic import BaseModel, Field


class Question(BaseModel):
    q_id: str = Field(min_length=1, description="Question or part id in the short form '1a', '2bii', '3' (number, then "
                                                "letter, then roman numerals; no 'Q', spaces or brackets)")
    text: str = Field(default="", description="The full question text for this part")
    max_marks: int = Field(ge=0, description="Marks shown on the paper for this part (0 if absent)")


class MarkPoint(BaseModel):
    label: str = Field(description="Allocation label as in the scheme, e.g. 'M1', 'A1', 'B1'")
    marks: int = Field(ge=0, description="Marks this allocation is worth")


class MarkSchemeEntry(BaseModel):
    q_id: str = Field(min_length=1, description="Question or part id this row marks, matching the paper's q_id")
    answer: str = Field(default="", description="The expected answer or method")
    marks: List[MarkPoint] = Field(default_factory=list, description="Mark allocations in scheme order")
    notes: str = Field(default="", description="Accept / reject notes, alternatives, ECF")


class Band(BaseModel):
    band: str = Field(description="Band name, e.g. 'A' or '4'")
    marks: int = Field(ge=0, description="Marks awarded for this band")
    descriptor: str = Field(default="", description="What an answer in this band looks like")


class RubricCriterionBands(BaseModel):
    criterion: str = Field(min_length=1, description="Criterion name, e.g. 'Organisation'")
    bands: List[Band] = Field(default_factory=list, description="Bands from best to worst")


_Q_ID = re.compile(r"^(q)?(\d+)([a-z])?([ivx]+)?$", re.IGNORECASE)
_ROMAN = re.compile(r"^[ivx]+$")
_QID_NOISE = re.compile(r"[\s().]+")


def norm_qid(q_id: Any) -> str:
    """Matching key for a question id, so the extractor's "Q1(a)" / "1 (a)" and the scheme's "1a" meet:
    lowercase, a leading "q" before a digit dropped, spaces, parentheses and dots removed. Every place that
    looks an extracted question up by a scheme row's q_id compares these keys, never the raw strings."""
    key = _QID_NOISE.sub("", str(q_id or "")).lower()
    if len(key) > 1 and key[0] == "q" and key[1].isdigit():
        key = key[1:]
    return key


def q_label(q_id: str) -> str:
    """Teacher-facing label for a question id: "1a" -> "1(a)", "2bii" -> "2(b)(ii)", "4iii" -> "4(iii)",
    "q1" -> "Q1". Anything else (already formatted, free text) is returned unchanged."""
    raw = (q_id or "").strip()
    m = _Q_ID.match(raw)
    if not m:
        return raw
    prefix, number, letter, roman = m.groups()
    suffix = (letter or "") + (roman or "")
    if len(suffix) > 1 and _ROMAN.match(suffix.lower()):
        # "4iii": the whole suffix is one roman part, not the letter "i" plus "ii".
        parts = [suffix.lower()]
    else:
        parts = [p.lower() for p in (letter, roman) if p]
    return ("Q" if prefix else "") + number + "".join(f"({p})" for p in parts)


def _get(item: Any, key: str, default: Any = None) -> Any:
    return item.get(key, default) if isinstance(item, dict) else getattr(item, key, default)


def scheme_total(kind: str, questions: Iterable[Union[Question, dict]], scheme: Iterable[Any]) -> int:
    """Total marks an assignment is out of. mark_scheme: the allocation total of each question's
    scheme row, falling back to the question's max_marks when it has no row (scheme rows alone when
    there are no questions). rubric: the best band of every criterion. criteria: the questions' marks."""
    questions = list(questions)
    scheme = list(scheme)
    if kind == "rubric":
        return sum(max((int(_get(b, "marks", 0)) for b in _get(c, "bands", []) or []), default=0) for c in scheme)
    if kind == "mark_scheme":
        row_totals = {_get(s, "q_id"): sum(int(_get(m, "marks", 0)) for m in _get(s, "marks", []) or []) for s in scheme}
        if not questions:
            return sum(row_totals.values())
        return sum(row_totals.get(_get(q, "q_id"), int(_get(q, "max_marks", 0))) for q in questions)
    return sum(int(_get(q, "max_marks", 0)) for q in questions)


# --- extraction agent IO ----------------------------------------------------------------------

import instructor  # noqa: E402
from atomic_agents import BaseIOSchema  # noqa: E402


class PaperExtractInput(BaseIOSchema):
    """Input to the paper extractor: the question paper's pages and a short hint about the paper."""

    images: List[instructor.Image] = Field(..., description="Question paper pages as images, in order")
    hint: str = Field(default="", description="Assignment title / subject, e.g. 'Sec 4 Maths — Quadratics'")


class PaperExtract(BaseIOSchema):
    """Paper extractor output: every question and part on the paper, in order."""

    questions: List[Question] = Field(default_factory=list, description="Questions and parts in paper order")


class SchemeExtractInput(BaseIOSchema):
    """Input to the scheme extractor: the mark scheme / rubric pages and the paper's question list."""

    images: List[instructor.Image] = Field(..., description="Mark scheme or rubric pages as images, in order")
    questions: List[Question] = Field(default_factory=list, description="The paper's questions, so rows align by q_id")


class SchemeExtract(BaseIOSchema):
    """Scheme extractor output for a mark scheme: one row per question part."""

    items: List[MarkSchemeEntry] = Field(default_factory=list, description="Mark scheme rows in scheme order")


class RubricExtract(BaseIOSchema):
    """Scheme extractor output for a rubric: one entry per criterion with its bands."""

    items: List[RubricCriterionBands] = Field(default_factory=list, description="Criteria with their bands")
